### ISSUE TITLE
Predictable Transaction ID Possible Vulnerability Allows Unauthorized Termination of OCPP Sessions

### ISSUE THREAD

**Issue by Tano-Coppoletta** [steve-community/steve#1296]
#### Checklist

- [x] I checked other issues already, but found no answer/solution
- [ ] I checked the documentation and wiki, but found no answer/solution
- [ ] I am running the latest version and the issue still occurs
- [x] I am sure that this issue is about SteVe (and not about the charging station software or something unrelated to SteVe)

#### Specifications
```
SteVe Version     : 3.6.0
Operating system  : Ubuntu 22.04.1
JDK               : openjdk version "11.0.20.1"
Database          : version 1.0.3
```

#### Docker Container Setup
Two Docker containers are set up to simulate two OCPP clients for testing purposes.

#### Issue Description
A possible security vulnerability was identified where unauthorized termination of transactions is possible due to predictable transaction Ids. The server issues a new transaction ID upon receiving an incomplete `StartTransaction` request and, due to the IDs being auto-incremented, an attacker can predict and use them to terminate other transactions.

#### Steps to Reproduce
1. From the first Docker container (Client 1), a transaction is started:
   ```json
   [2,"1","StartTransaction", {
       "connectorId": 1,
       "idTag": "test",
       "meterStart": 20,
       "timestamp": "2023-09-03T12:34:56Z"
   }]
	```
2. From the second Docker container (Client 2), a `StartTransaction` request with an empty idTag is sent to obtain a new transaction Id:
  ```json
     [2, "1", "StartTransaction", {"idTag": ""}]
  ```

3. Client 2 then sends a `StopTransaction` request using the new transaction ID minus 1:
  ```json
[2, "1","StopTransaction", {
    "meterStop": 1,
    "timestamp": "2023-11-06T15:54:23Z",
    "transactionId": [Retrieved Transaction ID - 1],
    "reason": "EVDisconnected",
    "transactionData": [...]
}]
  ```
4.The unauthorized termination of Client 1's transaction is confirmed via the web interface.
![Screenshot from 2023-11-08 16-30-39](https://github.com/steve-community/steve/assets/62875720/49fac839-9325-4928-8409-c0fef8f16d69)

### Expected behavior
The server should not issue a transaction ID for incomplete `StartTransaction` requests and should authenticate `StopTransaction` requests before processing.

### Actual behavior
The server processes unauthorized `StopTransaction` requests using predicted transaction IDs.

**Comment by lategoodbye** [comment]
From my point of view does the StartTransaction from point 2 violates the OCPP 1.6 schema (fields connectorId, meterStart and timestamp are missing). I think for OCPP 1.6J this should generate a CallError with ErrorCode = OccurenceConstraintViolation. There is no point in processing such a message, because no charging station should send such a mess.

**Comment by Tano-Coppoletta** [comment]
I appreciate your insight regarding the schema violation for the StartTransaction message. However, it's important to consider scenarios where a charging station might be compromised. If an attacker gains physical access to a charging station, they could potentially send malformed or incomplete OCPP messages. Therefore, server-side validation is crucial to safeguard against such threats.

**Comment by lategoodbye** [comment]
@Tano-Coppoletta Maybe you got me wrong, i wanted to say that Steve (server side) should generate a CallError and prevent further processing of the transaction. So your concerns are valid from my point of view.

**Comment by Tano-Coppoletta** [comment]
Hello,

I'm reaching out for an update regarding this issue.

We are in the process of submitting Common Vulnerabilities and Exposures (CVEs) related to this matter. However, we can wait if you are in the process of fixing it.

Additionally, we intend to publish our findings in an academic paper.

Your prompt response and attention to this matter would be greatly appreciated as it will greatly inform our next steps.

Thank you and looking forward to your response.

**Comment by goekay** [comment]
steve was _lenient_ when it comes to ocpp message strictness/validation this far (due to many charging station bad behaviour). however, recently https://github.com/steve-community/steve/pull/1953 enabled ocpp message validation which was added to [this](https://github.com/steve-community/steve/releases/tag/steve-3.11.0) release. therefore,

@lategoodbye 's argument

> From my point of view does the StartTransaction from point 2 violates the OCPP 1.6 schema (fields connectorId, meterStart and timestamp are missing). I think for OCPP 1.6J this should generate a CallError with ErrorCode = OccurenceConstraintViolation. There is no point in processing such a message, because no charging station should send such a mess.

and @Tano-Coppoletta 's argument 

> The server should not issue a transaction ID for incomplete StartTransaction requests [...]

should be taken care of. 

however, the second part of @Tano-Coppoletta 's argument 

>  [...] should authenticate StopTransaction requests before processing.

is still not executed, because of which i am leaving this issue as open.

**Comment by goekay** [comment]
notes about the untreated part:

> [...] should authenticate StopTransaction requests before processing.

we could logically check `meterStop` and `timestamp` from stop message against the respective values from start message, such that
- `stop.meterStop >= start.meterStart`
- `stop.timestamp >= start.timestamp`

and not update the database/transaction if violated (and still return a normal `StopTransaction.conf` as if everything was good). however, this would not close the attack vector absolutely, since the compromised station (or its impersonator) can still send unlimited number of StopTransactions (until giving up) with different values with the hope that _some_ will pass the validation and stop the transaction with faked values.

**Comment by juherr** [comment]
@goekay I think one current limitation of SteVe is that events and state are mixed. 

In this case, the StopTransaction should be accepted and stored as an event, but ignored at the state level (or the transaction should be flagged) if it violates business rules.

What do you think?

**Comment by goekay** [comment]
this is already happening to a degree: we have `transaction_start`, `transaction_stop`, and `transaction_stop_failed` tables. `transaction_start` and `transaction_stop` are self-explanatory. `transaction_stop_failed` is being used whenever a StopTransaction cannot be processed (e.g. transactionId cannot be found). then, we have the `transaction` database view which joins `transaction_start` and `transaction_stop` to create the final picture of a transaction. 

what part of this structure is a limitation?

we can flag the entry in `transaction_stop` according to the business rules and do the join of `transaction` with the latest _good_ stop entry.

edit: or, we could store the bad/flagged stop directly to `transaction_stop_failed`.

**Comment by juherr** [comment]
True, I missed some parts of the design. It should be possible to flag bad star, stop and maybe meters. 👍

**Comment by goekay** [comment]
@juherr dedicated issues for stricter StartTransaction and MeterValues processing will come later. or, feel free to create yourself.

**Comment by goekay** [comment]
i want to emphasize the following part of my previous comment:

> however, this would not close the attack vector absolutely, since the compromised station (or its impersonator) can still send unlimited number of StopTransactions (until giving up) with different values with the hope that some will pass the validation and stop the transaction with faked values.

i have no solution for this. an attacker can still set very high values for meterStop and stop timestamp, and it would pass the validation. 

stop timestamp case could be addressed by making sure that it is not in the future (maybe some additional delta to account for time differences between station's now and backend's now). 

on the other hand, meterStop is still an open case.

**Comment by goekay** [comment]
resolved by https://github.com/steve-community/steve/pull/1961 with the exception of tighter `meterStop` check.
